forge/test/mlir/llama/test_llama_inference.py

```diff
@@ -120,3 +121,346 @@ def test_llama_inference_cache_cpu():
+        model_inputs.append(
+            [
+                past_key_values[idx : idx + 2]
+                for idx in range(0, len(past_key_values), 2)
+            ]
+        )
+        generated_tokens = torch.cat(
+            [generated_tokens, next_token.unsqueeze(0)], dim=-1
+        )
+
+    # Generated text
+    generated_text = tokenizer.decode(generated_tokens[0], skip_special_tokens=True)
+    print("generated_text=", generated_text)
+
+
+# The test is just workaround for checking with/without padding.
+@pytest.mark.xfail()
+def test_llama_prefill_on_cpu_decode_on_tt_cache_padding():
+
+    # Load Llama 3B model and tokenizer
+    model_path = "openlm-research/open_llama_3b"
+    framework_model, tokenizer = load_model_and_tokenizer(
+        model_path=model_path, use_cache=True, return_dict=True
+    )
+
+    class LlamaModelWrapper(torch.nn.Module):
+        """
+        In LlamaModelWrapper class, forward function takes single input token (i.e last predicted token)
+        and past key values from the previous iteration and return logits and past key values
+        Args:
+            input_id (`torch.Tensor`) - shape of (batch_size, 1)
+            past_key_values (`List[List[torch.Tensor, torch.Tensor]]`) - key/values shape - (batch_size, num_of_key_values_heads, key_value_seq_len, head_dim)
+        Returns:
+            outputs - Logits of shape (batch_size, 1, vocab_size) and past_key_values (`List[List[torch.Tensor, torch.Tensor]]`)
+            past key/values tensor shape - (batch_size, num_of_key_values_heads, key_value_seq_len + 1, head_dim)
+        """
+
+        def __init__(self, model):
+            super().__init__()
+            self.model = model
+
+        def forward(self, input_id, past_key_values):
+            model_outputs = self.model(
+                input_ids=input_id, past_key_values=past_key_values
+            )
+            return model_outputs.logits, model_outputs.past_key_values
+
+    llama_model = LlamaModelWrapper(framework_model)
+
+    # Prepare input sentence
+    prompt = "Q: What is the largest animal?\nA:"
+    inputs = tokenizer(prompt, return_tensors="pt")
+
+    # Run Prefill on CPU with cache to get the initial logits and past key-values
+    prefill_output = framework_model(
+        input_ids=inputs.input_ids, attention_mask=inputs.attention_mask
+    )
+    next_token_logits = prefill_output.logits[:, -1, :]
+    next_token = torch.argmax(next_token_logits, dim=-1)
+
+    generated_tokens = inputs.input_ids
+    generated_tokens = torch.cat([generated_tokens, next_token.unsqueeze(0)], dim=-1)
+
+    past_key_values_list = [[k, v] for k, v in prefill_output.past_key_values]
+
+    model_inputs = [next_token.unsqueeze(0), past_key_values_list]
+
+    # Padding on past key values can be enabled by setting apply_padding_on_past_key_values to True.
+    apply_padding_on_past_key_values = True
+    max_new_tokens = 32
+    if apply_padding_on_past_key_values:
```

Comment:
@dgolubovicTT  We are only applying zero padding in past key values in key_value_seq_length dimension before passing to the decoder stage input.